week4/spiral/vitcra/Main.hs

```diff
@@ -0,0 +1,45 @@
+module Main where
+
+import           Data.List
+import           Data.Set  (Set)
+import qualified Data.Set  as S
+
+-- | Returns the graph set of the spiral,
+-- | i.e the points that are on the spiral.
+spiralGraph :: Int -> Set (Int, Int)
+spiralGraph n = fst $ foldl' f (S.empty, (0,-1)) $ zip counts deltas
+      -- | The start is outside the grid  ^^^^
+      -- | We'll step into (0,0) right away.
+  where
+    -- | Counts are basically the lengths of the lines of the spiral
+    -- | Here's the pattern: (n: pattern)
+    -- | 9: 9 9 8 7 6 5 4 3 2 1
+    -- | 8: 8 8 7 6 5 4 3 2 1
+    -- | etc
+    counts = n : [n, n-1..1]
+    -- | Go right, down, left, up and then cycle that.
+    deltas = cycle [(0,1), (1, 0), (0, -1), (-1, 0)]
+    -- | So zip counts deltas is like:
+    -- | go 9 times right, 8 times down, 8 times left, 6 times up, etc
+    -- | Put the visited points into the Graph set.
+
+    -- | Helper function that add two pairs
+    add (a, b) (c, d) = (a+c, b+d)
+    -- | The fold function that carries current position along with the Graph.
+    f (set, current) (count, delta) = foldl' g (set, current) [1..count] where
+      g (s, c) _ = let next = add c delta in (S.insert next s, next)
```

Comment:
Do you actually need the set in the fold? If all you do is insert elements into it, you can just produce a list of positions and do `Set.fromList` to it when you're done.